FirebaseRemoteConfig/Sources/FIRRemoteConfig.m

```diff
@@ -237,6 +240,61 @@ - (void)callListeners:(NSString *)key config:(NSDictionary *)config {
   }
 }
 
+- (void)setCustomSignals:(nullable NSDictionary<NSString *, NSObject *> *)customSignals
+          WithCompletion:(void (^_Nullable)(NSError *_Nullable error))completionHandler {
+  void (^setCustomSignalsBlock)(void) = ^{
+    if (!customSignals) {
+      if (completionHandler) {
+        dispatch_async(dispatch_get_main_queue(), ^{
+          completionHandler(nil);
+        });
+      }
+      return;
+    }
+
+    for (NSString *key in customSignals) {
+      NSObject *value = customSignals[key];
+      if (value && ![value isKindOfClass:[NSString class]] &&
+          ![value isKindOfClass:[NSNumber class]]) {
+        if (completionHandler) {
+          dispatch_async(dispatch_get_main_queue(), ^{
+            NSError *error =
+                [NSError errorWithDomain:FIRRemoteConfigCustomSignalsErrorDomain
+                                    code:FIRRemoteConfigCustomSignalsErrorInvalidValueType
+                                userInfo:@{
+                                  NSLocalizedDescriptionKey :
+                                      @"Invalid value type. Must be NSString or NSNumber."
+                                }];
+            completionHandler(error);
+          });
+        }
+        return;
+      }
+    }
+
+    NSMutableDictionary<NSString *, NSObject *> *newCustomSignals =
+        [[NSMutableDictionary alloc] initWithDictionary:self->_settings.customSignals];
+
+    for (NSString *key in customSignals) {
+      NSObject *value = customSignals[key];
+      if (value) {
+        [newCustomSignals setObject:value forKey:key];
+      } else {
+        [newCustomSignals removeObjectForKey:key];
+      }
+    }
+
+    self->_settings.customSignals = newCustomSignals;
```

Comment:
We execute setCustomSignalsBlock using dispatch_async(_queue, setCustomSignalsBlock). Since _queue is a serial queue for the read and write lock, it handles synchronization, ensuring that updates are processed in the correct order.